File: bot-example/kotlin/src/main/kotlin/PathFinding.kt
class PathFinding {
    companion object {
        fun solve(
            maze: Api.Options.Maze,
            start: Api.Options.CellPos,
            goal: Api.Options.CellPos
        ): List<Api.Options.CellPos> {
            val queue = ArrayDeque<Api.Options.CellPos>()
            val visited = mutableSetOf<Api.Options.CellPos>()
            val prev = mutableMapOf<Api.Options.CellPos, Api.Options.CellPos>()

            queue.add(start)
            visited.add(start)

            while (queue.isNotEmpty()) {
                val current = queue.removeFirst()
                if (current == goal) {
                    val path = mutableListOf<Api.Options.CellPos>()

                    var cur = current
                    while (cur != start) {
                        path.add(cur)
                        cur = prev[cur]!!
                    }
                    path.add(cur)
                    path.reverse()
                    return path
                }

                for (next in getNeighbors(maze, current)) {
                    if (next !in visited) {
                        queue.add(next)
                        visited.add(next)
                        prev[next] = current
                    }
                }
            }

            return emptyList()
        }

        private fun getNeighbors(maze: Api.Options.Maze, current: Api.Options.CellPos): List<Api.Options.CellPos> {
            val width = maze.width
            val height = maze.height
            val x = current.x
            val y = current.y

            val neighbors = mutableListOf<Api.Options.CellPos>()
            if (x > 0 && maze.isWall(x - 1, y).not()) {
                neighbors.add(Api.Options.CellPos(x - 1, y))
            }
            if (x < width - 1 && maze.isWall(x + 1, y).not()) {
                neighbors.add(Api.Options.CellPos(x + 1, y))
            }
            if (y > 0 && maze.isWall(x, y - 1).not()) {
                neighbors.add(Api.Options.CellPos(x, y - 1))
            }
            if (y < height - 1 && maze.isWall(x, y + 1).not()) {
                neighbors.add(Api.Options.CellPos(x, y + 1))
            }

            return neighbors
        }
    }
}

File: bot-example/csharp/PathFinding.cs
using CSharp.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharp
{
    public class PathFinding
    {
        public static List<CellPosition> Solve(Maze maze, CellPosition start, CellPosition goal)
        {
            var queue = new Queue<CellPosition>();
            var visited = new HashSet<CellPosition>();
            var prev = new Dictionary<CellPosition, CellPosition>();

            queue.Enqueue(start);
            visited.Add(start);

            var path = new List<CellPosition>();

            while (queue.Count > 0)
            {
                var current = queue.Dequeue();

                if (current == goal)
                {
                    while (current != start)
                    {
                        path.Add(current);
                        current = prev[current];
                    }
                    path.Add(current);
                    path.Reverse();
                    return path;
                }

                foreach (var next in GetNeighbors(maze, current))
                {
                    if (!visited.Contains(next))
                    {
                        queue.Enqueue(next);
                        visited.Add(next);
                        prev.Add(next, current);
                    }
                }
            }

            return path;
        }

        private static List<CellPosition> GetNeighbors(Maze maze, CellPosition current)
        {
            var width = maze.Width;
            var height = maze.Height;
            var x = current.X;
            var y = current.Y;

            var neighbors = new List<CellPosition>(4);

            if (x > 0 && !maze.IsWall(x - 1, y))
            {
                neighbors.Add(new CellPosition(x - 1, y));
            }
            if (x < width - 1 && !maze.IsWall(x + 1, y))
            {
                neighbors.Add(new CellPosition(x + 1, y));
            }
            if (y > 0 && !maze.IsWall(x, y - 1))
            {
                neighbors.Add(new CellPosition(x, y - 1));
            }
            if (y < height - 1 && !maze.IsWall(x, y + 1))
            {
                neighbors.Add(new CellPosition(x, y + 1));
            }

            return neighbors;
        }
    }
}
